File: src/bindings/ruby/test/TestFunctionDefinition.rb
#
# @file    TestFunctionDefinition.rb
# @brief   SBML FunctionDefinition unit tests
# @author  Akiya Jouraku (Ruby conversion)
# @author  Ben Bornstein 
#
# $Id$
# $HeadURL$
#
# This test file was converted from src/sbml/test/TestFunctionDefinition.c
# with the help of conversion sciprt (ctest_converter.pl).
#
#<!---------------------------------------------------------------------------
# This file is part of libSBML.  Please visit http://sbml.org for more
# information about SBML, and the latest version of libSBML.
#
# Copyright 2005-2008 California Institute of Technology.
# Copyright 2002-2005 California Institute of Technology and
#                     Japan Science and Technology Corporation.
# 
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation.  A copy of the license agreement is provided
# in the file named "LICENSE.txt" included with this software distribution
# and also available online as http://sbml.org/software/libsbml/license.html
#--------------------------------------------------------------------------->*/
#
require 'test/unit'
require 'libSBML'

class TestFunctionDefinition < Test::Unit::TestCase

  def setup
    @@fd = LibSBML::FunctionDefinition.new
    if (@@fd == nil)
    end
  end

  def teardown
    @@fd = nil
  end

  def test_FunctionDefinition_create
    assert( @@fd.getTypeCode == LibSBML::SBML_FUNCTION_DEFINITION )
    assert( @@fd.getMetaId == "" )
    assert( @@fd.getNotes == nil )
    assert( @@fd.getAnnotation == nil )
    assert( @@fd.getId == "" )
    assert( @@fd.getName == "" )
    assert( @@fd.getMath == nil )
  end

  def test_FunctionDefinition_createWith
    math = LibSBML::parseFormula("lambda(x, x^3)")
    fd = LibSBML::FunctionDefinition.new("pow3",math)
    assert( fd.getTypeCode == LibSBML::SBML_FUNCTION_DEFINITION )
    assert( fd.getMetaId == "" )
    assert( fd.getNotes == nil )
    assert( fd.getAnnotation == nil )
    assert( fd.getName == "" )
    math1 = fd.getMath
    assert( math1 != nil )
    formula = LibSBML::formulaToString(math1)
    assert( formula != nil )
    assert ((  "lambda(x, x^3)" == formula ))
    assert( fd.getMath != math )
    assert_equal true, fd.isSetMath
    assert ((  "pow3" == fd.getId ))
    assert_equal true, fd.isSetId
    math = nil
    fd = nil
  end

  def test_FunctionDefinition_free_NULL
    
  end

  def test_FunctionDefinition_getArguments
    @@fd.setMath(LibSBML::parseFormula("lambda(x, y, x^y)"))
    assert( @@fd.getNumArguments == 2 )
    math = @@fd.getArgument(0)
    assert( math != nil )
    assert_equal true, math.isName
    assert ((  "x" == math.getName ))
    assert( math.getNumChildren == 0 )
    math = @@fd.getArgument(1)
    assert( math != nil )
    assert_equal true, math.isName
    assert ((  "y" == math.getName ))
    assert( math.getNumChildren == 0 )
    assert( @@fd.getArgument(0) == @@fd.getArgument( "x") )
    assert( @@fd.getArgument(1) == @@fd.getArgument( "y") )
  end

  def test_FunctionDefinition_getBody
    math1 = LibSBML::parseFormula("lambda(x, x)")
    @@fd.setMath(math1)
    math = @@fd.getBody
    assert( math != nil )
    assert_equal true, math.isName
    assert ((  "x" == math.getName ))
    assert( math.getNumChildren == 0 )
    math1 = nil
  end

  def test_FunctionDefinition_setId
    id = "pow3"
    @@fd.setId(id)
    assert (( id == @@fd.getId ))
    assert_equal true, @@fd.isSetId
    if (@@fd.getId == id)
    end
    @@fd.setId(@@fd.getId)
    assert (( id == @@fd.getId ))
    @@fd.setId("")
    assert_equal false, @@fd.isSetId
    if (@@fd.getId != nil)
    end
  end

  def test_FunctionDefinition_setMath
    math = LibSBML::parseFormula("lambda(x, x^3)")
    @@fd.setMath(math)
    math1 = @@fd.getMath
    assert( math1 != nil )
    formula = LibSBML::formulaToString(math1)
    assert( formula != nil )
    assert ((  "lambda(x, x^3)" == formula ))
    assert( @@fd.getMath != math )
    assert_equal true, @@fd.isSetMath
    @@fd.setMath(@@fd.getMath)
    math1 = @@fd.getMath
    assert( math1 != nil )
    formula = LibSBML::formulaToString(math1)
    assert( formula != nil )
    assert ((  "lambda(x, x^3)" == formula ))
    assert( @@fd.getMath != math )
    @@fd.setMath(nil)
    assert_equal false, @@fd.isSetMath
    if (@@fd.getMath != nil)
    end
  end

  def test_FunctionDefinition_setName
    name = "Cube Me"
    @@fd.setName(name)
    assert (( name == @@fd.getName ))
    assert_equal true, @@fd.isSetName
    if (@@fd.getName == name)
    end
    @@fd.setName(@@fd.getName)
    assert (( name == @@fd.getName ))
    @@fd.setName("")
    assert_equal false, @@fd.isSetName
    if (@@fd.getName != nil)
    end
  end

end

File: src/TestFunctionDefinition.c
/**
 * Filename    : TestFunctionDefinition.c
 * Description : SBML FunctionDefinition unit tests
 * Author(s)   : SBML Development Group <sysbio-team@caltech.edu>
 * Organization: JST ERATO Kitano Symbiotic Systems Project
 * Created     : 2003-05-03
 * Revision    : $Id$
 * Source      : $Source$
 *
 * Copyright 2003 California Institute of Technology and
 * Japan Science and Technology Corporation.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
 * documentation provided hereunder is on an "as is" basis, and the
 * California Institute of Technology and Japan Science and Technology
 * Corporation have no obligations to provide maintenance, support,
 * updates, enhancements or modifications.  In no event shall the
 * California Institute of Technology or the Japan Science and Technology
 * Corporation be liable to any party for direct, indirect, special,
 * incidental or consequential damages, including lost profits, arising
 * out of the use of this software and its documentation, even if the
 * California Institute of Technology and/or Japan Science and Technology
 * Corporation have been advised of the possibility of such damage.  See
 * the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * The original code contained here was initially developed by:
 *
 *     Ben Bornstein
 *     The Systems Biology Markup Language Development Group
 *     ERATO Kitano Symbiotic Systems Project
 *     Control and Dynamical Systems, MC 107-81
 *     California Institute of Technology
 *     Pasadena, CA, 91125, USA
 *
 *     http://www.cds.caltech.edu/erato
 *     mailto:sysbio-team@caltech.edu
 *
 * Contributor(s):
 */


#include "sbml/common.h"
#include "sbml/FormulaParser.h"
#include "sbml/FunctionDefinition.h"


static FunctionDefinition_t *FD;


void
FunctionDefinitionTest_setup (void)
{
  FD = FunctionDefinition_create();

  if (FD == NULL)
  {
    fail("FunctionDefinition_create() returned a NULL pointer.");
  }
}


void
FunctionDefinitionTest_teardown (void)
{
  FunctionDefinition_free(FD);
}


START_TEST (test_FunctionDefinition_create)
{
  fail_unless( FD->typecode   == SBML_FUNCTION_DEFINITION, NULL );
  fail_unless( FD->metaid     == NULL, NULL );
  fail_unless( FD->notes      == NULL, NULL );
  fail_unless( FD->annotation == NULL, NULL );
  fail_unless( FD->id         == NULL, NULL );
  fail_unless( FD->name       == NULL, NULL );
  fail_unless( FD->math       == NULL, NULL );
}
END_TEST


START_TEST (test_FunctionDefinition_createWith)
{
  ASTNode_t            *math = SBML_parseFormula("lambda(x, x^3)");
  FunctionDefinition_t *fd   = FunctionDefinition_createWith("pow3", math);


  fail_unless( fd->typecode   == SBML_FUNCTION_DEFINITION, NULL );
  fail_unless( fd->metaid     == NULL, NULL );
  fail_unless( fd->notes      == NULL, NULL );
  fail_unless( fd->annotation == NULL, NULL );
  fail_unless( fd->name       == NULL, NULL );

  fail_unless( fd->math == math, NULL );
  fail_unless( FunctionDefinition_isSetMath(fd), NULL );

  fail_unless( !strcmp(fd->id, "pow3"), NULL );
  fail_unless( FunctionDefinition_isSetId(fd), NULL );

  FunctionDefinition_free(fd);
}
END_TEST


START_TEST (test_FunctionDefinition_free_NULL)
{
  FunctionDefinition_free(NULL);
}
END_TEST


START_TEST (test_FunctionDefinition_setId)
{
  char *id = "pow3";


  FunctionDefinition_setId(FD, id);

  fail_unless( !strcmp(FD->id, id), NULL );
  fail_unless( FunctionDefinition_isSetId(FD), NULL );

  if (FD->id == id)
  {
    fail("FunctionDefinition_setId(...) did not make a copy of string.");
  }

  /* Reflexive case (pathological) */
  FunctionDefinition_setId(FD, FD->id);
  fail_unless( !strcmp(FD->id, id), NULL );

  FunctionDefinition_setId(FD, NULL);
  fail_unless( !FunctionDefinition_isSetId(FD), NULL );

  if (FD->id != NULL)
  {
    fail("FunctionDefinition_setId(FD, NULL) did not clear string.");
  }
}
END_TEST


START_TEST (test_FunctionDefinition_setName)
{
  char *name = "Cube Me";


  FunctionDefinition_setName(FD, name);

  fail_unless( !strcmp(FD->name, name), NULL );
  fail_unless( FunctionDefinition_isSetName(FD), NULL );

  if (FD->name == name)
  {
    fail("FunctionDefinition_setName(...) did not make a copy of string.");
  }

  /* Reflexive case (pathological) */
  FunctionDefinition_setName(FD, FD->name);
  fail_unless( !strcmp(FD->name, name), NULL );

  FunctionDefinition_setName(FD, NULL);
  fail_unless( !FunctionDefinition_isSetName(FD), NULL );

  if (FD->name != NULL)
  {
    fail("FunctionDefinition_setName(FD, NULL) did not clear string.");
  }
}
END_TEST


START_TEST (test_FunctionDefinition_setMath)
{
  ASTNode_t *math = SBML_parseFormula("lambda(x, x^3)");


  FunctionDefinition_setMath(FD, math);

  fail_unless( FD->math == math, NULL );
  fail_unless( FunctionDefinition_isSetMath(FD), NULL );

  /* Reflexive case (pathological) */
  FunctionDefinition_setMath(FD, FD->math);
  fail_unless( FD->math == math, NULL );

  FunctionDefinition_setMath(FD, NULL);
  fail_unless( !FunctionDefinition_isSetMath(FD), NULL );

  if (FD->math != NULL)
  {
    fail("FunctionDefinition_setMath(FD, NULL) did not clear ASTNode.");
  }
}
END_TEST


Suite *
create_suite_FunctionDefinition (void)
{
  Suite *suite = suite_create("FunctionDefinition");
  TCase *tcase = tcase_create("FunctionDefinition");


  tcase_add_checked_fixture( tcase,
                             FunctionDefinitionTest_setup,
                             FunctionDefinitionTest_teardown );

  tcase_add_test( tcase, test_FunctionDefinition_create     );
  tcase_add_test( tcase, test_FunctionDefinition_createWith );
  tcase_add_test( tcase, test_FunctionDefinition_free_NULL  );
  tcase_add_test( tcase, test_FunctionDefinition_setId      );
  tcase_add_test( tcase, test_FunctionDefinition_setName    );
  tcase_add_test( tcase, test_FunctionDefinition_setMath    );

  suite_add_tcase(suite, tcase);

  return suite;
}
